watch-resource action, for live updating dashboards

malcolmholmes opened the pull request:
Grafana 7.3 includes (behind a feature flag) support for live updating of saved dashboards. When a dashboard is saved somewhere, all open dashboards are reloaded.

This feature piggy-backs on that: when invoked, Grizzly connects to a web socket on the remote Grafana instance. It then waits. When a dashboard is saved, it receives a notification, retrieves the dashboard JSON and writes it to a local file.

For a dashboard with a UID `useful-dashboard` Syntax could be:

```
grr watch-resource dashboard.useful-dashboard dashboard.json
```

The principal usecase for this feature is when working on a dashboard that is managed in version control. If we first use `grr apply` to push our dashboard up to Grafana, then use `grr watch-resource` to spot changes, we can now effectively use Grafana itself as a dashboard IDE - dashboard saves are written directly back to the file system where they can be committed back to version control.

Still to do:

- [x] Give useful error message if Grafana instance doesn't have the ws feature flag enabled
- [x] Are all of the event types needed? They were copied wholesale from example code - review is needed.
- [x] Stop listening (end program) if subscription ends, or Grafana disappears
- [ ] Document the feature

Beyond this, given that this feature is not jsonnet specific, it could easily be a part of a more generic `grafanactl` tool, that Grizzly could consume. Grizzly could (relatively) easily be split into a Jsonnet processor and a `grafanactl` tool that could focus simply on lifecycle of static resources.

sdboyer commented on pkg/grizzly/workflow.go:
```diff
@@ -316,6 +316,40 @@ func Watch(config Config, watchDir string, parser Parser) error {
+		resourceID = parts[2]
+
+	} else {
+		return fmt.Errorf("UID must be <provider>.<uid>: %s", UID)
```
This seems inconsistent with the two possible branches above - could be two OR three parts. Is this an artifact of some old form of our dashboard UIDs?

malcolmholmes replied:
Before when Grizzly just handled dashboards, the UID was just the dashboard UID. Now we need to 'namespace' the UID, so we might call it `dashboard.my-dash`. But what if Grizzly might support more than one dashboard provider? Then we would prefix it with a provider name: `grafana.dashboard.my-dash`. That's the idea, anyway. Maybe overkill. And the error is unhelpful and out of date - it is really the 'object type' or 'kind' or such, not the 'provider'. I need to think about that.

sdboyer reviewed (commented):
Looks like a pretty good first pass to me! Couple questions/comments. Makes sense that you'd want to complete the items in the description before merging, though. Do you want (more) review before then?

malcolmholmes commented on cmd/grr/workflow.go:
```diff
@@ -128,6 +128,20 @@ func watchCmd(config grizzly.Config) *cli.Command {
+func watchResourceCmd(config grizzly.Config) *cli.Command {
+	cmd := &cli.Command{
+		Use:   "watch-resource <uid-to-watch> <output-file>",
+		Short: "watch for file changes on remove and save locally",
```
ahh "remote", fixing...

malcolmholmes commented on cmd/grr/workflow.go:
```diff
@@ -128,6 +128,20 @@ func watchCmd(config grizzly.Config) *cli.Command {
+		Args:  cli.ArgsExact(2),
+	}
+	cmd.Run = func(cmd *cli.Command, args []string) error {
+		uid := args[0]
```
that is absolutely possible. At the moment, Grafana offers a per dashboard channel. @ryantxu has already suggested that we could have a channel to listen for changes on *any* dashboard, which would make this sort of feature possible. Then we could as you say listen to all dashboards, or a regexp, or just one folder, etc.

ryantxu replied:
:+1:

malcolmholmes commented on pkg/grafana/dashboard-watcher.go:
Laziness? I didn't see the point in defining a type that was only going to be used once! Saves me one or two lines of code!

sdboyer replied:
It's not defining a type any more than what you're doing now - it's the same anonymous struct as you're using. Just, using a `var` decl (which results in the same empty default value being assigned to `response`) instead of `:=`

malcolmholmes replied:
Ahh, gotcha. No idea. Tend to use := rather than var, prefer it. Not aware of a reason when to choose one over the other.

malcolmholmes commented on pkg/grafana/dashboard-watcher.go:
yup, removed.

malcolmholmes commented:
Upon suggestion from @Duologic , I have renamed `watch-resource` to `listen`. Much simpler.

I've also reduced the number of events listened to, and improved the messages from the ones that remain. And, I've added code to make Grizzly shut down if Grafana disappears. I couldn't make it react to the disappearance of the dashboard. Need to investigate that more.

malcolmholmes commented:
This PR is now ready for review/merge. I notice the README is totally out of date. I'll update the readme and add `listen` in a separate PR.
